### PR TITLE
SB-713: `StorageProvider` refactored to use `java.nio.file.Path` abstraction

### PR DESCRIPTION
**Pull request by sbespalov**

### CODE REVIEW

**Comment by carlspring** [comment]
@cs-devops-bot rebuild.

**Comment by carlspring** [comment]
@cs-devops-bot rebuild.

**Comment by carlspring** [comment]
@cs-devops-bot rebuild.

**Comment by carlspring** [comment]
@cs-devops-bot add to whitelist

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java [hunk 1]
```diff
@@ -1,87 +1,69 @@
...
 
-    /**
-     * This is to only be used internally in case the artifact coordinates have not been passed in.
-     */
-    private String relativePath;
-
+    private String baseDir;
 
-    public ArtifactFile(Repository repository,
+    public ArtifactFile(String baseDir,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java]: Please, rename this to `repositoryBaseDir`.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java [hunk 2]
```diff
@@ -1,87 +1,69 @@
...
-    public ArtifactFile(Repository repository,
-                        String path)
-    {
-        this(repository, path, false);
+        this(baseDir, artifactCoordinates, false);
     }
 
-    public ArtifactFile(Repository repository,
-                        String path,
+    public ArtifactFile(String baseDir,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java]: Please, rename this to `repositoryBaseDir`.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java (hunk 1 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java]: done
Note that `Temp` logic moved to `RepositoryFileSystem`

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java (hunk 2 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactFile.java]: done
Note that `Temp` logic moved to `RepositoryFileSystem`

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 3]
```diff
@@ -63,4 +79,132 @@ public Storage getStorage(String storageId)
...
+    
+    protected boolean isTrash(String path){
+        return path.contains(".trash");
+    }
+    
+    protected boolean isTemp(String path){
+        return path.contains(".temp");
+    }
+
+    protected boolean isIntex(String path){
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: The name of this method has a spelling mistake. It's supposed to be `isIndex`.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 4]
```diff
@@ -63,4 +79,132 @@ public Storage getStorage(String storageId)
...
+    {
+        Storage storage = getConfiguration().getStorage(storageId);
+        Repository repository = storage.getRepository(repositoryId);
+        StorageProvider storageProvider = storageProviderRegistry.getProvider(repository.getImplementation());
+        OutputStream os;
+        if (isArtifact(repository, path, false))
+        {
+            ArtifactPath artifactPath = resolve(repository, getArtifactCoordinates(path));
+            os = storageProvider.getOutputStreamImplementation(artifactPath);
+        } else
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: Cosmetic: Could you please move the `else` on a new line?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 5]
```diff
@@ -63,4 +79,132 @@ public Storage getStorage(String storageId)
...
+        }
+
+        return os instanceof ArtifactOutputStream ? (ArtifactOutputStream) os : new ArtifactOutputStream(os, null);
+    }
+    
+    protected abstract boolean isMetadata(String path);
+    
+    protected abstract boolean isChecksum(String path);
+    
+    protected boolean isTrash(String path){
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: Cosmetic: Could you please sort out all the curly braces to match the project's coding convention and be on the next line? (Could you please have a look at the rest of your code for occurrences of  `) {` and sort them out...?). Thanks! :)

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java (hunk 4 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: done

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java (hunk 5 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: done

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java (hunk 3 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: done

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactOutputStream.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import org.carlspring.strongbox.artifact.coordinates.ArtifactCoordinates;
+
+/**
+ * This OutputStream wraps a source stream from different Storage types (File System, AWS, JDBC, etc.).
+ * 
+ * @author Sergey Bespalov
+ *
+ */
+public class ArtifactOutputStream
+        extends FilterOutputStream
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactOutputStream.java]: This class should extend `MultipleDigestOutputStream` instead, as it should be able to produce checksums.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java [hunk 7]
```diff
@@ -464,28 +264,25 @@ public void deleteMetadata(String storageId,
 
         Storage storage = getConfiguration().getStorage(storageId);
         Repository repository = storage.getRepository(repositoryId);
-        StorageProvider storageProvider = storageProviderRegistry.getProvider(repository.getImplementation());
-
-        final File repoPath = storageProvider.getFileImplementation(repository.getBasedir());
-
+        RepositoryPath repositoryPath = resolve(repository);
+        if (!Files.isDirectory(repositoryPath))
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: I'm afraid I don't quite agree with this change. What you're doing in this case is changing the logic/behaviour, as far as I can see. My code is specifically for directories, whereas this is ignoring directories, as far as I understand.

 Could you please do a diff with my original code and make sure this is right, or explain what this is correcting, if I'm misunderstanding you...? Thanks! :)

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java [hunk 8]
```diff
@@ -60,40 +52,15 @@ public NugetHierarchicalArtifactCoordinates getArtifactCoordinates(String path)
...
     {
-        Storage storage = getConfiguration().getStorage(storageId);
-        Repository repository = storage.getRepository(repositoryId);
-        
-        NugetHierarchicalArtifactCoordinates coordinates = new NugetHierarchicalArtifactCoordinates(path);
-        ArtifactFile artifactFile = new ArtifactFile(repository, coordinates);
-        artifactFile.createParents();
-
-        return new ArtifactFileOutputStream(artifactFile);
+        return path.endsWith("nupkg.sha512");
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java]: I believe this should actually just be:

    return path.endsWith(".sha512");

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java [hunk 9]
```diff
@@ -99,7 +101,17 @@ public ArtifactInputStream getInputStream(String storageId,
...
             {
-                final ArtifactInputStream is = resolveArtifact(sId, r.getId(), artifactPath);
+                ArtifactInputStream is;
+                try
+                {
+                    is = resolveArtifact(sId, r.getId(), artifactPath);
+                }
+                catch (FileNotFoundException e)
+                {
+                    logger.debug(String.format("Artifact not found in repository, continue group artifact search: repo-[%s]; path-[%s].",
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: We don't need this extra output. If it's not found, that's absolutely okay. If this is reported by Sonar, please, ignore the warning.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java (hunk 7 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/Maven2LayoutProvider.java]: If `repositoryPath` is not a directory it just do nothing and returns. So it is specifically for directories as originally.

File: strongbox-storage/strongbox-storage-indexing/pom.xml [hunk 10]
```diff
@@ -247,6 +247,17 @@
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
         </dependency>
+
```
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-indexing/pom.xml]: you are right, it was by accident. will be removed

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java (hunk 8 above)
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java]: Actually, we have a class called `ArtifactFileUtils`. I think the `isChecksum` method in it should simply be changed to support and {{SHA}} algorithm via a regex. We could then use that method instead. :)

File: strongbox-storage/strongbox-storage-indexing/pom.xml (hunk 10 above)
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/pom.xml]: Great, thanks! :)

Comment (sbespalov) [inline strongbox-storage/strongbox-storage-indexing/pom.xml]: done

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java (hunk 9 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java]: done

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java (hunk 8 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java]: done

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactOutputStream.java (hunk 6 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/io/ArtifactOutputStream.java]: done

**Pull request merged**